datadog_checks_base/datadog_checks/base/utils/tls.py

```diff
@@ -0,0 +1,139 @@
+}
+
+
+class TlsContextWrapper(object):
+    __slots__ = ('logger', 'config', 'tls_context')
+
+    def __init__(self, instance, remapper=None):
+        # type: (InstanceType, Dict[AnyStr, Dict[AnyStr, Any]]) -> None
+        default_fields = dict(STANDARD_FIELDS)
+
+        # Populate with the default values
+        config = {field: instance.get(field, value) for field, value in iteritems(default_fields)}
+        for field in STANDARD_FIELDS:
+            unique_name = UNIQUE_FIELD_PREFIX + field
+            if unique_name in instance:
+                config[unique_name] = instance[unique_name]
+
+        if remapper is None:
+            remapper = {}
+
+        for remapped_field, data in iteritems(remapper):
+            field = data.get('name')
+
+            if field.startswith(UNIQUE_FIELD_PREFIX):
+                standard_field_name = field[len(UNIQUE_FIELD_PREFIX) :]
+            else:
+                standard_field_name = field
+
+            # Ignore fields we don't recognize
+            if standard_field_name not in STANDARD_FIELDS:
+                continue
+
+            # Ignore remapped fields if the standard one is already used
+            if field in instance:
+                continue
+
+            # Invert default booleans if need be
+            default = default_fields[standard_field_name]
+            if data.get('invert'):
+                default = not default
+
+            # Get value, with a possible default
+            value = instance.get(remapped_field, data.get('default', default))
+
+            # Invert booleans if need be
+            if data.get('invert'):
+                value = not is_affirmative(value)
+
+            config[field] = value
+
+        if any(
+            (config['tls_ca_cert'], config['tls_cert'], config['tls_private_key'], config['tls_private_key_password'])
+        ):
+            config['tls_verify'] = True
+
+        # Populate with the higher-priority configuration options if set
+        for field in default_fields:
+            unique_name = UNIQUE_FIELD_PREFIX + field
+            if unique_name in config:
+                config[field] = config[unique_name]
+                del config[unique_name]
+
+        self.config = config
+        self.tls_context = self._create_tls_context()
+
+    def _create_tls_context(self):
+        # type: () -> ssl.SSLContext
+
+        # https://docs.python.org/3/library/ssl.html#ssl.SSLContext
+        # https://docs.python.org/3/library/ssl.html#ssl.PROTOCOL_TLS
+        context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLS)
+
+        # https://docs.python.org/3/library/ssl.html#ssl.SSLContext.verify_mode
+        if not self.config['tls_verify']:
+            context.verify_mode = ssl.CERT_NONE
+            return context
+        context.verify_mode = ssl.CERT_REQUIRED
+
+        # https://docs.python.org/3/library/ssl.html#ssl.SSLContext.check_hostname
+        context.check_hostname = self.config['tls_validate_hostname']
```

Comment:
I think we should switch the order: https://github.com/DataDog/integrations-core/blob/a25cecca0d34f51a8066509d5479fee6a54e01c7/tls/datadog_checks/tls/tls.py#L348-L353

Reply:
That's an interesting catch. We aren't affected by that new behavior (because when we set `check_hostname`, `verify_mode` is necessary `CERT_REQUIRED` already.
But I can switch the order and point to the documentation accordingly so that we prevent unexpected bugs in the future.

Reply:
Actually thinking about that, I can't switch the order.
On python2 and python3 < 3.7, setting `check_hostname` to True before `verify_mode` raises an Exception.
So setting `verify_mode` first to CERT_REQUIRED then setting `check_hostname` to the value of the config seems to be the only way.